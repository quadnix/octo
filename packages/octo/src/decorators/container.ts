import type { Constructable } from '../app.type.js';

type Factory<T> = { create: (...args: unknown[]) => Promise<T> };
type FactoryValue<T> = Factory<T> | [Promise<Factory<T>>, (factory: Factory<T>) => void];

/**
 * The Container class is a box to hold all registered factories.
 *
 * In Octo, we minimize the use of the `new` keyword and get instance of a class using its factory.
 * The factory has full control over how the instance is created.
 *
 * By using the Container and Factory concepts, it is possible to override internal class definitions at runtime.
 * It is an incredibly powerful tool to customize implementation.
 */
export class Container {
  private static FACTORY_TIMEOUT_IN_MS = 5000;

  private static readonly factories: {
    [key: string]: {
      default: boolean;
      factory: FactoryValue<any>;
      metadata: { [key: string]: string };
    }[];
  } = {};

  /**
   * `Container.get()` allows to get an instance of a class using its factory.
   *
   * @param type The type or name of the class to get.
   * @param options Allows selection of a specific factory to use to get the instance.
   * - The `args` option supplies arguments to the factory.
   * - The `metadata` option identifies the factory.
   * @returns The instance of the class.
   */
  static async get<T>(
    type: Constructable<T> | string,
    options?: {
      args?: unknown[];
      metadata?: { [key: string]: string };
    },
  ): Promise<T> {
    const args = options?.args || [];
    const metadata = options?.metadata;
    const name = typeof type === 'string' ? type : type.name;

    if (!(name in this.factories)) {
      this.factories[name] = [];
    }

    let factoryContainer: (typeof Container.factories)[keyof typeof Container.factories][0] | undefined;
    if (!metadata) {
      factoryContainer = this.factories[name].find((f) => f.default);
    } else {
      const filters: { key: string; value: string }[] = [];
      for (const key in metadata) {
        filters.push({ key, value: metadata[key] });
      }
      factoryContainer = this.factories[name].find((f) => filters.every((c) => f.metadata[c.key] === c.value));
    }

    if (factoryContainer?.factory) {
      if (!Array.isArray(factoryContainer.factory)) {
        return (factoryContainer.factory as Factory<T>).create(...args);
      } else {
        const factory = await (factoryContainer.factory[0] as Promise<Factory<T>>);
        return factory.create(...args);
      }
    }

    const newFactoryContainer = {
      default: true,
      metadata,
    } as (typeof Container.factories)[keyof typeof Container.factories][0];
    this.factories[name].push(newFactoryContainer);

    let promiseResolver;
    let promiseTimeout;
    const promise = new Promise<Factory<T>>((resolve, reject) => {
      promiseTimeout = setTimeout(() => {
        reject(new Error(`Timed out waiting for factory ${name} to resolve!`));
      }, this.FACTORY_TIMEOUT_IN_MS);
      promiseResolver = resolve;
    });
    newFactoryContainer.factory = [promise, promiseResolver];

    const factory = await promise;
    if (promiseTimeout) {
      clearTimeout(promiseTimeout);
    }
    return factory.create(...args);
  }

  /**
   * `Container.registerFactory()` allows to register a factory for a class.
   *
   * @param type The type or name of the class for which the factory is registered.
   * @param factory The factory class being registered.
   * @param options Distinguishes between different factories of the same class.
   * - The `metadata` attaches custom metadata to the factory.
   */
  static registerFactory<T>(
    type: Constructable<T> | string,
    factory: Factory<T>,
    options?: {
      metadata?: { [key: string]: string };
    },
  ): void {
    const metadata = options?.metadata || {};
    const name = typeof type === 'string' ? type : type.name;

    if (!(name in this.factories)) {
      this.factories[name] = [];
    }

    const filters: { key: string; value: string }[] = [];
    for (const key in metadata) {
      filters.push({ key, value: metadata[key] });
    }
    const factoryContainer = this.factories[name].find((f) => filters.every((c) => f.metadata[c.key] === c.value));

    if (factoryContainer?.factory) {
      // If factory is not a promise set by get() above, it has already been registered.
      if (!Array.isArray(factoryContainer.factory)) {
        return;
      }

      factoryContainer.metadata = metadata;
      const resolve = factoryContainer.factory[1] as (factory: Factory<T>) => void;
      resolve(factory);
      factoryContainer.factory = factory;
      this.setDefault(type, factory);

      return;
    }

    this.factories[name].push({ default: false, factory, metadata });
    this.setDefault(type, factory);
  }

  /**
   * `Container.reset()` clears all registered factories and empties the container. This is mostly used in testing.
   */
  static reset(): void {
    this.FACTORY_TIMEOUT_IN_MS = 5000;

    for (const name in this.factories) {
      delete this.factories[name];
    }
  }

  /**
   * `Container.setDefault()` sets a default factory for the class.
   *
   * @example
   * ```ts
   * // Register a factory.
   * Container.registerFactory(MyClass, MyClassFactory, { metadata: { key: 'value' } });
   *
   * // Without default.
   * Container.get(MyClass, { metadata: { key: 'value' } });
   *
   * // Set default.
   * Container.setDefault(MyClass, MyClassFactory);
   *
   * // With default.
   * Container.get(MyClass);
   * ```
   * @param type The type or name of the class for which the default factory is set.
   * @param factory The factory class being set as default.
   */
  static setDefault<T>(type: Constructable<T> | string, factory: Factory<T>): void {
    const name = typeof type === 'string' ? type : type.name;

    this.factories[name]?.forEach((f) => {
      f.default = (f.factory as unknown as Constructable<T>).name === (factory as unknown as Constructable<T>).name;
    });
  }

  /**
   * The Container, by default, will wait a maximum of 5 seconds for a factory to resolve and provide an instance.
   * This method allows to change that timeout.
   *
   * @param timeoutInMs The timeout in milliseconds.
   */
  static setFactoryTimeout(timeoutInMs: number): void {
    this.FACTORY_TIMEOUT_IN_MS = timeoutInMs;
  }

  /**
   * `Container.unRegisterFactory()` will unregister all factories of a class.
   *
   * @param type The type or name of the class for which all factory is unregistered.
   */
  static unRegisterFactory<T>(type: Constructable<T> | string): void {
    const name = typeof type === 'string' ? type : type.name;
    delete this.factories[name];
  }
}
