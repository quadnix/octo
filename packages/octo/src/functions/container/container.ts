import type { Constructable } from '../../app.type.js';

type Factory<T> = { create: (...args: unknown[]) => Promise<T> };
type FactoryValue<T> = Factory<T> | [Promise<Factory<T>>, (factory: Factory<T>) => void];

/**
 * The Container class is a box to hold all registered factories.
 *
 * In Octo, we minimize the use of the `new` keyword and get instance of a class using its factory.
 * The factory has full control over how the instance is created.
 *
 * By using the Container and Factory concepts, it is possible to override internal class definitions at runtime.
 * It is an incredibly powerful tool to customize implementation.
 */
export class Container {
  private static FACTORY_TIMEOUT_IN_MS = 5000;

  private static readonly factories: { [key: string]: { factory: FactoryValue<any> } } = {};

  /**
   * `Container.get()` allows to get an instance of a class using its factory.
   *
   * @param type The type or name of the class to get.
   * @param options Allows selection of a specific factory to use to get the instance.
   * - The `args` option supplies arguments to the factory.
   * @returns The instance of the class.
   */
  static async get<T>(
    type: Constructable<T> | string,
    options?: {
      args?: unknown[];
    },
  ): Promise<T> {
    const args = options?.args || [];
    const name = typeof type === 'string' ? type : type.name;

    const factoryContainer: (typeof Container.factories)[keyof typeof Container.factories] | undefined =
      this.factories[name];
    if (factoryContainer?.factory) {
      if (!Array.isArray(factoryContainer.factory)) {
        return (factoryContainer.factory as Factory<T>).create(...args);
      } else {
        const factory = await (factoryContainer.factory[0] as Promise<Factory<T>>);
        return factory.create(...args);
      }
    }

    const newFactoryContainer = {} as (typeof Container.factories)[keyof typeof Container.factories];
    this.factories[name] = newFactoryContainer;

    let promiseResolver;
    let promiseTimeout;
    const promise = new Promise<Factory<T>>((resolve, reject) => {
      promiseTimeout = setTimeout(() => {
        reject(new Error(`Timed out waiting for factory ${name} to resolve!`));
      }, this.FACTORY_TIMEOUT_IN_MS);
      promiseResolver = resolve;
    });
    newFactoryContainer.factory = [promise, promiseResolver];

    const factory = await promise;
    if (promiseTimeout) {
      clearTimeout(promiseTimeout);
    }
    return factory.create(...args);
  }

  /**
   * `Container.registerFactory()` allows to register a factory for a class.
   *
   * @param type The type or name of the class for which the factory is registered.
   * @param factory The factory class being registered.
   */
  static registerFactory<T>(type: Constructable<T> | string, factory: Factory<T>): void {
    const name = typeof type === 'string' ? type : type.name;

    const factoryContainer = this.factories[name];
    if (factoryContainer?.factory) {
      // If factory is not a promise set by get() above, it has already been registered.
      if (!Array.isArray(factoryContainer.factory)) {
        return;
      }

      const resolve = factoryContainer.factory[1] as (factory: Factory<T>) => void;
      resolve(factory);
      factoryContainer.factory = factory;

      return;
    } else {
      this.factories[name] = { factory };
    }
  }

  /**
   * `Container.reset()` clears all registered factories and empties the container. This is mostly used in testing.
   */
  static reset(): void {
    this.FACTORY_TIMEOUT_IN_MS = 5000;

    for (const name in this.factories) {
      delete this.factories[name];
    }
  }

  /**
   * The Container, by default, will wait a maximum of 5 seconds for a factory to resolve and provide an instance.
   * This method allows to change that timeout.
   *
   * @param timeoutInMs The timeout in milliseconds.
   */
  static setFactoryTimeout(timeoutInMs: number): void {
    this.FACTORY_TIMEOUT_IN_MS = timeoutInMs;
  }

  /**
   * `Container.unRegisterFactory()` will unregister all factories of a class.
   *
   * @param type The type or name of the class for which all factory is unregistered.
   */
  static unRegisterFactory<T>(type: Constructable<T> | string): void {
    const name = typeof type === 'string' ? type : type.name;
    delete this.factories[name];
  }

  static async waitToResolveAllFactories(): Promise<void> {
    const promiseToResolveAllFactories: Promise<Factory<unknown>>[] = [];

    for (const name of Object.keys(this.factories)) {
      const factoryContainer = this.factories[name];
      if (factoryContainer?.factory && Array.isArray(factoryContainer.factory)) {
        promiseToResolveAllFactories.push(factoryContainer.factory[0]);
      }
    }

    await Promise.all(promiseToResolveAllFactories);
  }
}
