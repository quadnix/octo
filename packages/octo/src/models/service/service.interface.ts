import { Service } from './service.model.js';

export interface IService {
  serviceId: Service['serviceId'];
}
