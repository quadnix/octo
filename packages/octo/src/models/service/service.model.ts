import { Model } from '../../decorators/model.decorator.js';
import { ModelError } from '../../errors/index.js';
import { AModel } from '../model.abstract.js';
import { ServiceSchema } from './service.schema.js';

/**
 * A Service model can be any third-party service that your application relies on.
 * What separates them from Server is that Services are hosted and managed outside of your infrastructure.
 * An external Redis server on the cloud, or a managed Kafka queue from AWS, are a few examples.
 *
 * @example
 * ```ts
 * const service = new Service('MyService');
 * ```
 * @group Models
 * @see Definition of [Default Models](/docs/fundamentals/models#default-models).
 */
@Model<Service>('@octo', 'service', ServiceSchema)
export class Service extends AModel<ServiceSchema, Service> {
  readonly serviceId: string;

  constructor(serviceId: string) {
    super();
    this.serviceId = serviceId;
  }

  override setContext(): string {
    const parents = this.getParents();
    const app = parents['app'][0].to;
    return [`${(this.constructor as typeof Service).NODE_NAME}=${this.serviceId}`, app.getContext()].join(',');
  }

  override synth(): ServiceSchema {
    throw new ModelError('Method not implemented! Use subclass', this);
  }
}
