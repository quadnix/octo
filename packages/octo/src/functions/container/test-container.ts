import type { Constructable } from '../../app.type.js';
import { DiffUtility } from '../diff/diff.utility.js';
import { Container } from './container.js';

type Factory<T> = { create: () => Promise<T> };

type FactoryMock<T> = {
  metadata?: { [key: string]: string };
  type: Constructable<T> | string;
  value: T;
};

type TestContainerSubjects = {
  importFrom?: Constructable<IPackageMock>[];
  mocks?: FactoryMock<unknown>[];
};
type TestContainerOptions = { factoryTimeoutInMs?: number };

export interface IPackageMock {
  destroy: () => Promise<void>;

  getMocks: () => FactoryMock<unknown>[];

  init(): Promise<void>;
}

/**
 * The TestContainer class is an isolated {@link Container} for tests.
 *
 * :::info Info
 * A test framework like Jest, supports parallel test executions in different cores of the machine.
 * Each test is a single test file, whose `it()` blocks are executed serially.
 * :::
 *
 * TestContainer is created with testing framework's parallel execution in mind.
 * A TestContainer should be created in the `beforeAll()` block of your test,
 * which modifies the Container for the duration of the test.
 * This TestContainer does not affect other tests in parallel since they are on a separate machine core.
 * Once tests are done executing, the `afterAll()` block cleans up the Container.
 */
export class TestContainer {
  private static packageMocks: IPackageMock[] = [];

  private static createTestFactory<T>(mock: FactoryMock<T>): Factory<T> {
    return class {
      static mock: FactoryMock<T> = mock;

      static async create(): Promise<T> {
        return this.mock.value;
      }
    };
  }

  /**
   * The `TestContainer.create()` method allows you to mock factories.
   *
   * @example
   * ```ts
   * beforeAll(async () => {
   *   await TestContainer.create({
   *     importFrom: [OctoAwsCdkPackageMock],
   *     mocks: [
   *       { type: MyClass, value: jest.fn() },
   *       { metadata: { key: 'value' }, type: AnotherClass, value: new AnotherClass() },
   *     ],
   *   }, { factoryTimeoutInMs: 500 });
   * });
   *
   * ```
   * @param subjects The subjects being mocked.
   * - `importFrom` is an array of {@link IPackageMock} classes, to import custom mocks,
   * such as from octo-aws-cdk package - [OctoAwsCdkPackageMock](/api/octo-aws-cdk/class/OctoAwsCdkPackageMock).
   * - `mocks` is an array of objects, to override the default factories.
   *   - Use `metadata?: { [key: string]: string }` to identify the factory being mocked.
   *   - Use `type: Constructable<T> | string` to identify the class being mocked.
   *   - Use `value: T` to provide the mocked value.
   * @param options Options to configure TestContainer.
   * - `factoryTimeoutInMs?: number` is to override the default container timeout.
   */
  static async create(subjects: TestContainerSubjects, options?: TestContainerOptions): Promise<void> {
    const oldFactories = { ...Container['factories'] };
    Container.reset();

    if (options?.factoryTimeoutInMs) {
      Container.setFactoryTimeout(options.factoryTimeoutInMs);
    }

    const importedMocks: FactoryMock<unknown>[] = [];
    for (const mockClass of subjects.importFrom || []) {
      const mock = new mockClass();
      await mock.init();

      importedMocks.push(...mock.getMocks());
      this.packageMocks.push(mock);
    }

    subjects.mocks = [...importedMocks, ...(subjects.mocks || [])];

    for (const mock of subjects.mocks) {
      const mockClassName = typeof mock.type === 'string' ? mock.type : mock.type.name;
      const mockMetadata = mock.metadata || {};
      const registeredFactories = oldFactories[mockClassName] || [];
      const matchingFactoryIndex = registeredFactories.findIndex((f) =>
        DiffUtility.isObjectDeepEquals(f.metadata, mockMetadata),
      );

      if (matchingFactoryIndex > -1) {
        oldFactories[mockClassName][matchingFactoryIndex].factory = this.createTestFactory(mock);
      }
    }

    for (const name in oldFactories) {
      for (const oldFactory of oldFactories[name]) {
        Container.registerFactory(name, oldFactory.factory as Factory<unknown>, { metadata: oldFactory.metadata });
        if (oldFactory.default) {
          Container.setDefault(name, oldFactory.factory as Factory<unknown>);
        }
      }
    }
  }

  /**
   * The `TestContainer.reset()` method will completely destroy all factories from the Container.
   * Because the entire Container is destroyed, and not just the mocks,
   * it must always be called in your `afterAll()` block.
   *
   * @example
   * ```ts
   * afterAll(async () => {
   *   await TestContainer.reset();
   * });
   * ```
   */
  static async reset(): Promise<void> {
    for (const mock of this.packageMocks) {
      await mock.destroy();
    }
    Container.reset();
  }
}
