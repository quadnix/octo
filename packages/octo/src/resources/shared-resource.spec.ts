import { jest } from '@jest/globals';
import { ActionInputs, ActionOutputs, UnknownResource } from '../app.type.js';
import { Container } from '../decorators/container.js';
import { Factory } from '../decorators/factory.decorator.js';
import { Resource } from '../decorators/resource.decorator.js';
import { Diff, DiffAction } from '../functions/diff/diff.model.js';
import { IAction } from '../models/action.interface.js';
import { App } from '../models/app/app.model.js';
import { Region } from '../models/region/region.model.js';
import { ResourceSerializationService } from '../services/serialization/resource/resource-serialization.service.js';
import { TransactionService } from '../services/transaction/transaction.service.js';
import { IResourceAction } from './resource-action.interface.js';
import { AResource } from './resource.abstract.js';
import { ASharedResource } from './shared-resource.abstract.js';

@Factory<TransactionService>(TransactionService, { metadata: { key: 'test' } })
class TransactionServiceTestFactory {
  static async create(): Promise<TransactionService> {
    return new TransactionService();
  }
}

@Resource()
class ParentResource extends AResource<ParentResource> {
  readonly MODEL_NAME: string = 'parent-resource';

  constructor(resourceId: string) {
    super(resourceId, {}, []);
  }
}

@Resource()
class TestResource extends AResource<TestResource> {
  readonly MODEL_NAME: string = 'test-resource';

  constructor(resourceId: string, properties: { [key: string]: string }, parents: [ParentResource]) {
    super(resourceId, properties, parents);
  }
}

@Resource()
class SharedTestResource extends ASharedResource<TestResource> {
  readonly MODEL_NAME: string = 'shared-test-resource';

  constructor(resourceId: string, properties: object, parents: [TestResource?]) {
    super(resourceId, {}, parents as AResource<TestResource>[]);
  }
}

const universalResourceAction: IResourceAction = {
  ACTION_NAME: 'universal',
  filter: () => true,
  handle: jest.fn() as jest.Mocked<any>,
};

describe('SharedResource UT', () => {
  let resourceSerializationService: ResourceSerializationService;
  let transactionService: TransactionService;

  beforeAll(async () => {
    Container.setDefault(TransactionService, TransactionServiceTestFactory);

    resourceSerializationService = await Container.get(ResourceSerializationService);
  });

  beforeEach(async () => {
    transactionService = await Container.get(TransactionService);
  });

  afterAll(() => {
    Container.reset();
  });

  it('should serialize and deserialize empty shared-resources', async () => {
    const serializedOutput = resourceSerializationService.serialize([]);
    expect(serializedOutput).toMatchInlineSnapshot(`
      {
        "dependencies": [],
        "resources": {},
        "sharedResources": {},
      }
    `);
  });

  it('should serialize and deserialize shared-resources with parent', async () => {
    const parentResource = new ParentResource('parent-1');
    const testResource = new TestResource('resource-1', { property1: 'property-value-1' }, [parentResource]);
    testResource.response['response1'] = 'response-value-1';
    const sharedTestResource = new SharedTestResource('shared-test-resource', {}, [testResource]);

    const serializedOutput = resourceSerializationService.serialize([parentResource, testResource, sharedTestResource]);
    expect(serializedOutput).toMatchSnapshot();

    const resources = await resourceSerializationService.deserialize(serializedOutput);

    expect(resources['parent-1'].resourceId).toBe('parent-1');
    expect(resources['parent-1'].properties).toEqual({});
    expect(resources['parent-1'].response).toEqual({});
    expect(resources['parent-1']['dependencies'].length).toBe(1);

    expect(resources['resource-1'].resourceId).toBe('resource-1');
    expect(resources['resource-1'].properties).toEqual({ property1: 'property-value-1' });
    expect(resources['resource-1'].response).toEqual({ response1: 'response-value-1' });
    expect(resources['resource-1']['dependencies'].length).toBe(2);

    expect(resources['shared-test-resource'].resourceId).toBe('shared-test-resource');
    expect(resources['shared-test-resource']['dependencies'].length).toBe(1);
  });

  it('should serialize and deserialize shared-resources with different parents', async () => {
    const app = new App('app');

    // Assume a region is being created.
    const region1 = new Region('region-1');
    app.addRegion(region1);
    const diffs1 = [new Diff(region1, DiffAction.ADD, 'regionId', 'region-1')];

    // The region-1 determined its resources will have a parent resource, and a shared resource parented by parent-1.
    const parentResource1 = new ParentResource('parent-1');
    const testResource1 = new TestResource('resource-1', { property1: 'property-value-1' }, [parentResource1]);
    const sharedTestResource1 = new SharedTestResource('shared-test-resource', {}, [testResource1]);

    const addRegion1Action: IAction<ActionInputs, ActionOutputs> = {
      ACTION_NAME: 'test1',
      collectInput: () => [],
      collectOutput: () => ['parent-1', 'resource-1', 'shared-test-resource'],
      filter: (diff: Diff) => {
        return diff.value === 'region-1';
      },
      handle: (jest.fn() as jest.Mocked<any>).mockReturnValue({
        'parent-1': parentResource1,
        'resource-1': testResource1,
        'shared-test-resource': sharedTestResource1,
      }),
      revert: jest.fn() as jest.Mocked<any>,
    };

    transactionService.registerModelActions([addRegion1Action]);
    transactionService.registerResourceActions([universalResourceAction]);
    const generator1 = transactionService.beginTransaction(diffs1, {}, {}, { yieldNewResources: true });

    // Validate shared resources properties, responses, and dependencies are serialized correctly.
    const newResources1 = await generator1.next();
    const serializedOutput1 = resourceSerializationService.serialize(newResources1.value as UnknownResource[]);
    expect(serializedOutput1).toMatchSnapshot();

    // Assume another region is created.
    const region2 = new Region('region-2');
    app.addRegion(region2);
    const diffs2 = [new Diff(region2, DiffAction.ADD, 'regionId', 'region-2')];

    // The region-2 determined its resources will have a parent resource, and a shared resource parented by parent-2.
    const parentResource2 = new ParentResource('parent-2');
    const testResource2 = new TestResource('resource-2', { property2: 'property-value-2' }, [parentResource2]);
    const sharedTestResource2 = new SharedTestResource('shared-test-resource', {}, [testResource2]);

    const addRegion2Action: IAction<ActionInputs, ActionOutputs> = {
      ACTION_NAME: 'test2',
      collectInput: () => [],
      collectOutput: () => ['parent-2', 'resource-2', 'shared-test-resource'],
      filter: (diff: Diff) => {
        return diff.value === 'region-2';
      },
      handle: (jest.fn() as jest.Mocked<any>).mockReturnValue({
        'parent-2': parentResource2,
        'resource-2': testResource2,
        'shared-test-resource': sharedTestResource2,
      }),
      revert: jest.fn() as jest.Mocked<any>,
    };

    transactionService.registerModelActions([addRegion2Action]);
    const generator2 = transactionService.beginTransaction(
      diffs2,
      {
        'parent-1': parentResource1,
        'resource-1': testResource1,
        'shared-test-resource': sharedTestResource1,
      },
      // copy of old resources.
      {
        'parent-1': parentResource1,
        'resource-1': testResource1,
        'shared-test-resource': sharedTestResource1,
      },
      {
        yieldNewResources: true,
      },
    );

    // Validate shared resources properties, responses, and dependencies are serialized correctly.
    const newResources2 = await generator2.next();
    const serializedOutput2 = resourceSerializationService.serialize(newResources2.value as UnknownResource[]);
    expect(serializedOutput2).toMatchSnapshot();
  });

  describe('merge()', () => {
    it('should merge old shared resource with self and return new shared resource', () => {
      const resource1 = new ParentResource('resource-1');
      const resource2 = new ParentResource('resource-2');

      const sharedResource1 = new SharedTestResource('shared-resource', {}, [resource1]);
      sharedResource1.properties['property-1'] = 'property-value-1';
      sharedResource1.response['response-1'] = 'response-value-1';

      const sharedResource2 = new SharedTestResource('shared-resource', {}, [resource2]);
      sharedResource2.properties['property-2'] = 'property-value-2';
      sharedResource2.response['response-2'] = 'response-value-2';

      const mergedSharedResource = sharedResource2.merge(sharedResource1);

      expect(mergedSharedResource['dependencies'].length).toBe(2);
      expect(mergedSharedResource.properties).toMatchInlineSnapshot(`
        {
          "property-1": "property-value-1",
          "property-2": "property-value-2",
        }
      `);
      expect(mergedSharedResource.response).toMatchInlineSnapshot(`
        {
          "response-1": "response-value-1",
          "response-2": "response-value-2",
        }
      `);
    });
  });
});
