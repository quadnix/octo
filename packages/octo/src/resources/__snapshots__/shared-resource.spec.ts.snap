// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedResource UT should serialize and deserialize shared-resources with different parents 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "parent-resource=parent-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "test-resource=resource-1",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "parent-resource=parent-1",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "shared-test-resource=shared-test-resource",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "shared-test-resource=shared-test-resource",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "test-resource=resource-1",
    },
  ],
  "resources": {
    "parent-1": {
      "className": "ParentResource",
      "resource": {
        "properties": {},
        "resourceId": "parent-1",
        "response": {},
      },
    },
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "property1": "property-value-1",
        },
        "resourceId": "resource-1",
        "response": {},
      },
    },
  },
  "sharedResources": {
    "shared-test-resource": {
      "className": "SharedTestResource",
      "resource": {
        "properties": {},
        "resourceId": "shared-test-resource",
        "response": {},
      },
    },
  },
}
`;

exports[`SharedResource UT should serialize and deserialize shared-resources with different parents 2`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "parent-resource=parent-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "test-resource=resource-1",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "parent-resource=parent-1",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "shared-test-resource=shared-test-resource",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "shared-test-resource=shared-test-resource",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "shared-test-resource=shared-test-resource",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "test-resource=resource-2",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "shared-test-resource=shared-test-resource",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "test-resource=resource-1",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "parent-resource=parent-2",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "test-resource=resource-2",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-2",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "parent-resource=parent-2",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-2",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "shared-test-resource=shared-test-resource",
    },
  ],
  "resources": {
    "parent-1": {
      "className": "ParentResource",
      "resource": {
        "properties": {},
        "resourceId": "parent-1",
        "response": {},
      },
    },
    "parent-2": {
      "className": "ParentResource",
      "resource": {
        "properties": {},
        "resourceId": "parent-2",
        "response": {},
      },
    },
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "property1": "property-value-1",
        },
        "resourceId": "resource-1",
        "response": {},
      },
    },
    "resource-2": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "property2": "property-value-2",
        },
        "resourceId": "resource-2",
        "response": {},
      },
    },
  },
  "sharedResources": {
    "shared-test-resource": {
      "className": "SharedTestResource",
      "resource": {
        "properties": {},
        "resourceId": "shared-test-resource",
        "response": {},
      },
    },
  },
}
`;

exports[`SharedResource UT should serialize and deserialize shared-resources with parent 1`] = `
{
  "dependencies": [
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "parent-resource=parent-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "test-resource=resource-1",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "parent-resource=parent-1",
    },
    {
      "behaviors": [
        {
          "forAction": "delete",
          "onAction": "delete",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "test-resource=resource-1",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "parent",
      },
      "to": "shared-test-resource=shared-test-resource",
    },
    {
      "behaviors": [
        {
          "forAction": "add",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
        {
          "forAction": "update",
          "onAction": "add",
          "onField": "resourceId",
          "toField": "resourceId",
        },
      ],
      "from": "shared-test-resource=shared-test-resource",
      "relationship": {
        "onField": "resourceId",
        "toField": "resourceId",
        "type": "child",
      },
      "to": "test-resource=resource-1",
    },
  ],
  "resources": {
    "parent-1": {
      "className": "ParentResource",
      "resource": {
        "properties": {},
        "resourceId": "parent-1",
        "response": {},
      },
    },
    "resource-1": {
      "className": "TestResource",
      "resource": {
        "properties": {
          "property1": "property-value-1",
        },
        "resourceId": "resource-1",
        "response": {
          "response1": "response-value-1",
        },
      },
    },
  },
  "sharedResources": {
    "shared-test-resource": {
      "className": "SharedTestResource",
      "resource": {
        "properties": {},
        "resourceId": "shared-test-resource",
        "response": {},
      },
    },
  },
}
`;
